### ISSUE TITLE
[rl] Autodiff graph built and discarded on every rollout step — add `forward_inner`/no-...

### ISSUE THREAD

**Issue by anthonytorlucci** [anthonytorlucci/rlevo#177]
- **Labels:** enhancement
- **Files:** `algorithms/ppo/policies/categorical.rs`, `algorithms/ppo/policies/gaussian.rs`, `algorithms/ppo/ppo_policy.rs`, `algorithms/ppo/ppo_value.rs`, `algorithms/ppg/policies/categorical.rs`, `algorithms/ppg/ppg_policy.rs`
- **Body:** Every PPO/PPG policy and value head runs `sample_with_logprob`/`forward` through the full autodiff-tracked backend during environment rollout, building and immediately discarding a computation graph on every single `act()` call — the single measurable-runtime-impact perf finding across the on-policy code. Fix: add an inner-backend `forward_inner`/no-grad snapshot path (mirroring the pattern already used correctly elsewhere in the crate, e.g. SAC) and route rollout collection through it. Source: `-rl-ppo.md`, `-rl-ppg.md`.
- **Severity:** No High, but 6 files converge on identical High-effort-worthy perf finding, each doc's own "top three."

**Comment by anthonytorlucci** [comment]
Additional file-specific detail from the 2026-06-30 review of `gaussian.rs` (`crates/rlevo-reinforcement-learning/src/algorithms/ppo/policies/gaussian.rs`), distinct from but compounding this issue's autodiff-graph claim:

`sample_with_logprob` (gaussian.rs:551-556) calls `self.mean(obs.clone())` directly, then calls `self.log_prob_entropy(obs, z.clone())`, which internally calls `self.mean(obs)` again (gaussian.rs:492) — the two-hidden-layer MLP forward runs twice per `act()` call, plus an extra `obs.clone()`. This is on top of (not a duplicate of) the forward_inner/no-grad-snapshot fix this issue tracks: even after routing rollout through an inner-backend forward, the duplicate `mean()` invocation inside `log_prob_entropy` would still double the matmul cost unless `sample_with_logprob` is refactored to compute `mean` once and thread it into a `log_prob_entropy_from_mean`-style helper. Still present in the current file; verified by direct read, not the stale line numbers in the original review draft.

**Comment by anthonytorlucci** [comment]
Related but distinct mechanism, confirmed current (2026-07-20) in `crates/rlevo-reinforcement-learning/src/algorithms/ppg/policies/categorical.rs:269-281`.

`PpgCategoricalPolicyHead::deterministic_env_row_inner` (the greedy/eval path, on `B::InnerBackend`) can't call `PpgAuxValueHead::logits` because that trait is bound to `AutodiffBackend` only, so it recomputes logits manually:

```rust
let logits = inner.logits_head.forward(inner.trunk(obs)); // categorical.rs:276
```

Correction to how an earlier review framed this: `inner: &Self::InnerModule` is `PpgCategoricalPolicyHead<B::InnerBackend>` — the *same* struct, just parameterized by the inner backend — so this is not a privacy violation (it's an inherent method called from the type's own `impl` block, which is ordinary Rust). The real defect is a genericity gap: because `logits()` is only defined for `AutodiffBackend`, *any* second implementer of `PpgAuxValueHead` would have to duplicate this same trunk+head recomputation by hand for its own inner-backend eval path, rather than reusing a trait-level `forward_inner`-style method. That's the same root cause this issue already tracks for the rollout/act() paths — same fix shape (inner-backend forward without autodiff), different call site (greedy eval vs. rollout sampling).
